### ISSUE TITLE
F2003 `Nonlabel_Do_Stmt` behaves differently for string and `get_reader(string)`

### ISSUE THREAD

**Issue by TeranIvy** [stfc/fparser#114]
Running fparser2 F2003 `Nonlabel_Do_Stmt` directly on the string
```
do_loop = "do i=1), 2+"
objt = Nonlabel_Do_Stmt(do_loop)
```
falls over nicely as it should with `NoMatchError: Nonlabel_Do_Stmt: 'do i=1), 2+'`.

However, `Nonlabel_Do_Stmt(get_reader(do_loop))` does not raise an error, as noticed in [PSyclone PR #215 comment](https://github.com/stfc/PSyclone/pull/215#issuecomment-426632188). `get_reader(do_loop)` returns `None` because of incorrect Fortran and `Nonlabel_Do_Stmt(get_reader(do_loop))` returns `NoneType` instead of falling over ([later comment](https://github.com/stfc/PSyclone/pull/215#issuecomment-426712192)).

**Comment by rupertford** [comment]
@TeranIvy. I've still not got my head round the passing of strings vs the passing of reader objects and the returning of NoneType instead of NoMatchError. I think strings tend to return one and reader objects the other but that is not very deep insight.

However, I don't think what you report is a bug. NoMatchError and NoneType both signal that there was no match. NoMatchError usually returning the class and string that did not match does not affect the FortranSyntaxError that is returned from the program.

I think the reason that class and string are returned is to avoid recursion issues, but I've not tested out whether this is actually ever an issue.

So in summary, I don't think this is a bug and I suggest we close the issue.

**Comment by hiker** [comment]
It is confusing if parsing the same code presented once as string, once via a get_reader wrapper results in two different behaviours - it certainly caused me to waste quite a bit of time. Imho the behaviour should be consistent - and that is raise an exception in case of an error (since details about the error can be communicated to the caller this way).
Shouldn't it be possible to simply test if the argument is a FortranStringReader, and if so, parse the string directly to that we can get an exception (and also the key question: what happens if the line is in a file??)

**Comment by rupertford** [comment]
Hi @hiker. I'm not in a position to fully answer your point but what I can say is ...

1) It is not necessarily an error, it the mechanism to determine that there is a failure to match for a particular rule. Only when all possible matches have failed can we say that it is a (syntax) error.

2) it is unclear to me whether we want to communicate why there has been a failure to match. Personally I lean towards yes, so that we can report more detailed information on a syntax error. However, on a syntax error, there will be many match failures and determining which one is the most likely culprit might not be easy (or possible in some cases).

3) the whole code is written in the same way, so any major changes to the way matches work will involve a lot of code modification. That's not to say we shouldn't do it, it's just that we (well certainly I) need to better understand what is done now and why it is done this way before deciding what to do.

4) I think the code works on the reader base class so the same thing should happen with a FortranStringReader or a FortranFileReader.

**Comment by rupertford** [comment]
Closing this issue as it is not something we are planning to look at in the near future and does not affect functionality.
